src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1125,6 +1140,88 @@ class FunctionBinaryArithmetic : public IFunction
+
+        return executeArray(arguments, result_type, input_rows_count);
+
+        // if (typeid_cast<const ColumnConst *>(arguments[0].column.get()))
+        //     return result_column;
+        // else
+    }
+
+    ColumnPtr executeArray(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const
+    {
+        if constexpr (is_multiply || is_division)
+            throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Cannot use multiplication or division on arrays");
+
+        auto num_args = arguments.size();
+        DataTypes data_types;
+
+        ColumnsWithTypeAndName new_arguments {num_args};
+        DataTypePtr t;
+
+        const auto * left_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        const auto * right_const = typeid_cast<const ColumnConst *>(arguments[1].column.get());
+
+        /// Unpacking arrays if both are constants.
+        if (left_const && right_const)
+        {
+            new_arguments[0] = {left_const->getDataColumnPtr(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {right_const->getDataColumnPtr(), arguments[1].type, arguments[1].name};
+            auto col = executeImpl(new_arguments, result_type, 1);
+            return ColumnConst::create(std::move(col), input_rows_count);
+        }
+
+        /// Unpacking arrays if at least one column is constant.
+        if (left_const || right_const)
+        {
+            new_arguments[0] = {arguments[0].column->convertToFullColumnIfConst(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {arguments[1].column->convertToFullColumnIfConst(), arguments[1].type, arguments[1].name};
+            return executeImpl(new_arguments, result_type, input_rows_count);
+        }
+
+        /// Unpacking non-const arrays and checking sizes of them.
+        if (*typeid_cast<const ColumnArray *>(arguments[0].column.get())->getOffsets().data() !=
+            *typeid_cast<const ColumnArray *>(arguments[1].column.get())->getOffsets().data())
+        {
+            throw Exception(ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH,
+            "Cannot apply operation for arguments of different sizes. Size of the first argument: {}, size of the second argument: {}",
+            *typeid_cast<const ColumnArray *>(arguments[0].column.get())->getOffsets().data(),
+            *typeid_cast<const ColumnArray *>(arguments[1].column.get())->getOffsets().data());
+        }
+
+        auto a = typeid_cast<const ColumnArray *>(arguments[0].column.get())->getData().getPtr();
+        t = typeid_cast<const DataTypeArray *>(arguments[0].type.get())->getNestedType();
+        new_arguments[0] = {a, t, arguments[0].name};
+
+        a = typeid_cast<const ColumnArray *>(arguments[1].column.get())->getData().getPtr();
+        t = typeid_cast<const DataTypeArray *>(arguments[1].type.get())->getNestedType();
+        new_arguments[1] = {a, t, arguments[1].name};
```

Comment:
Please reduce the `typeid_cast<const DataTypeArray *>` here. It makes the code really hard to read. Feel free to save the result of casts into variable so you don't have to carry out the cast every time.